File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0, max=0;
        LinkedList<LinkedList<Integer>> graph = new LinkedList<>();
        for(int i=0; i<=n; i++){
            graph.add(new LinkedList<Integer>());
        }
        
        // 간선 정보 넣기
        for(int i=0; i<edge.length; i++){
            int v1 = edge[i][0], v2 = edge[i][1];
            graph.get(v1).add(v2);
            graph.get(v2).add(v1);
        }
        
        Queue<Integer> pq = new LinkedList<>();
        int[] edges = new int[n+1];
        
        pq.offer(1);
        
        while(!pq.isEmpty()){
            int node = pq.poll();
            LinkedList<Integer> row = graph.get(node);
            for(int v: row){
                if(v!=1&&edges[v]==0){
                    edges[v]=edges[node]+1;
                    max = max>edges[v] ? max:edges[v];
                    pq.offer(v);
                }
            }
        }
        
        for(int i=1; i<n+1; i++){
            if(edges[i]==max) answer++;
        }
        
        return answer;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    let answer = 0, max = 0;
    let graph = Array.from(Array(n+1), ()=>[]);
    let nodes = new Array(n+1).fill(-1);
    let queue = [];
    
    // 그래프 정보 채우기
    for(let row of edge){
        let v1 = row[0], v2=row[1];
        graph[v1].push(v2);
        graph[v2].push(v1);
    }
    
    // 1번 노드 - 각 노드 사이의 간선 개수 구하기
    queue.push(1);
    nodes[1]=0;
    while(queue.length>0){
        let node = queue.shift();
        for(let v of graph[node]){
            if(nodes[v]===-1){
                nodes[v] = nodes[node]+1;
                if(max<nodes[v]){
                    max = nodes[v];
                    answer=0;
                }
                answer++;
                queue.push(v);
            }
        }
    }
    
    return answer;
}